### PR TITLE
S3E pooling of embeddings

### PR DESCRIPTION
**Pull request by tholor**
Implementing a new pooling method for embeddings as proposed by Wang et al in their paper "Efficient Sentence Embedding via Semantic Subspace Analysis".

It could be a useful alternative to sentence embeddings generated via naive reduce_mean etc.

- Paper:  https://arxiv.org/abs/2002.09620
- Code: https://github.com/BinWang28/Sentence-Embedding-S3E

### CODE REVIEW

**Comment by tholor** [comment]
How to use it? 
- call `fit_s3_on_corpus()` to fit s3e on your corpus (or a subset)
- save results or pass them directly into the Inferencer to get your sentence embeddings 
``` 
   inferencer = Inferencer(model=model, processor=processor, task_type="embeddings", gpu=use_gpu,
                       batch_size=batch_size, extraction_strategy="s3e", extraction_layer=-1,
                       s3e_stats=s3e_stats)
```
See example script (https://github.com/deepset-ai/FARM/blob/s3e_pooling/examples/embeddings_extraction_s3e_pooling.py)

Known limitations: 
- Speed of `fit_s3_on_corpus`: 
-- no multiprocessing for tokenization during 
-- the original implementation for removing PCA components is slow and was not optimized yet
- Minor differences to the original method (handling of UNK tokens)

File: farm/modeling/language_model.py [hunk 1]
```diff
@@ -1046,11 +1070,10 @@ def save(self, save_dir):
         :param save_dir: The directory in which the model should be saved.
         :type save_dir: str
         """
-        raise NotImplementedError
         #save model
-        # self.model.save(save_dir=save_dir)
-        # #save config
-        # self.save_config(save_dir=save_dir)
+        self.model.save(save_dir=save_dir)
```
Comment (Timoeller) [inline farm/modeling/language_model.py:1074]: For me this hangs for normal sized (2-3GB) Glove/Word2Vec embeddings. 
I am unsure why, could you test on your machine?

File: farm/modeling/tokenization.py [hunk 2]
```diff
@@ -125,6 +126,8 @@ def __init__(
         :param do_lower_case: Flag whether to lower case the input
         :type do_lower_case: bool
         """
+        # TODO check why EmbeddingTokenizer.tokenize gives many UNK, while tokenize_with_metadata() works fine
```
Comment (Timoeller) [inline farm/modeling/tokenization.py:129]: resolved?

File: test/samples/s3e/fitted_s3e/language_model_config.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "embeddings_filename": "vectors.txt",
+  "hidden_size": 300,
+  "language": "German",
+  "name": "WordEmbedding_LM",
+  "vocab_filename": "vocab.txt",
+  "vocab_size": 4008
```
Comment (Timoeller) [inline test/samples/s3e/fitted_s3e/language_model_config.json:7]: unsure if this is really used, but it should be 113

File: examples/embeddings_extraction_s3e_pooling.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    ]
+
+    # Get embeddings for input text
+    result = inferencer.inference_from_dicts(dicts=basic_texts)
+    print(result)
+
+
+if __name__ == "__main__":
+    #TODO update to public model / data
+    lang_model = Path("saved_models/s3e_fasttext")
```
Comment (Timoeller) [inline examples/embeddings_extraction_s3e_pooling.py:101]: I know we said we want to have the path relative to top folder farm, but we never actually did. So lets use ../savedmodel and ../data, right?

File: farm/modeling/tokenization.py (hunk 2 above)
Comment (tholor) [reply farm/modeling/tokenization.py:129]: Nope, this is still happening. As we only use tokenize_with_metadata in FARM this is not super critical, but we should tackle it at some point. I guess it's related to whitespace tokenization not working properly in the EmbeddingTokenizer

File: farm/modeling/language_model.py (hunk 1 above)
Comment (tholor) [reply farm/modeling/language_model.py:1074]: It's not hanging for me, but it's crazily slow (~ 15min for german glove). I added a progress bar now to make it more transparent. We should then very soon accelerate this in a separate PR - potentially using numpy.savetxt or changing the format to h5py / torch .bin ...

File: test/samples/s3e/fitted_s3e/language_model_config.json (hunk 3 above)
Comment (tholor) [reply test/samples/s3e/fitted_s3e/language_model_config.json]: Fixed

File: examples/embeddings_extraction_s3e_pooling.py (hunk 4 above)
Comment (tholor) [reply examples/embeddings_extraction_s3e_pooling.py]: Fixed

**Pull request merged**